File: lab3/zad1/cpp/twolist.cpp
#include <iostream>
#include <fstream>
#include "node.h"
#include "twolist.h"

using namespace std;

TwoList::TwoList() {
    head = nullptr;
    tail = nullptr;
}

TwoList::~TwoList() {
    TwoListNode* curr = head;
    while (curr != nullptr) {
        TwoListNode* next = curr->next;
        delete curr;
        curr = next;
    }
}

void TwoList::addToHead(const string& val) {
    auto *newNode = new TwoListNode(val);
    newNode->prev = nullptr;    
    if (head == nullptr) {
        newNode->next = nullptr;
        head = newNode;
        tail = newNode;
    } else {
        newNode->next = head;
        head->prev = newNode;
        head = newNode;
    }
}

void TwoList::addToTail(const string& val) {
    auto *newNode = new TwoListNode(val);
    newNode->next = nullptr;
    if (head == nullptr) {
        newNode->prev = nullptr;
        head = newNode;
        tail = newNode;
    } else {
        newNode->prev = tail;
        tail->next = newNode;
        tail = newNode;
    }
}

void TwoList::addAfterValue(const string& targetVal, const string& newVal) {
    TwoListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }

    auto *newNode = new TwoListNode(newVal);
    newNode->next = targetNode->next;
    newNode->prev = targetNode;
    
    if (targetNode->next != nullptr) {
        targetNode->next->prev = newNode;
    }
    targetNode->next = newNode;
    
    if (targetNode == tail) {
        tail = newNode;
    }
}

void TwoList::addBeforeValue(const string& targetVal, const string& newVal) {
    TwoListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }
    
    if (targetNode == head) {
        addToHead(newVal);
        return;
    }

    auto *newNode = new TwoListNode(newVal);
    newNode->next = targetNode;
    newNode->prev = targetNode->prev;
    
    if (targetNode->prev != nullptr) {
        targetNode->prev->next = newNode;
    }
    targetNode->prev = newNode;
}

void TwoList::removeFromHead() {
    if (head == nullptr) {
        throw runtime_error("error");
    }    
    TwoListNode* temp = head;    
    if (head == tail) {
        head = nullptr;
        tail = nullptr;
    } else {
        head = head->next;
        head->prev = nullptr;
    }    
    delete temp;
}

void TwoList::removeFromTail() {
    if (head == nullptr) {
        throw runtime_error("error");
    }
    TwoListNode* temp = tail;
    if (head == tail) {
        head = nullptr;
        tail = nullptr;
    } else {
        tail = tail->prev;
        tail->next = nullptr;
    }
    delete temp;
}

void TwoList::removeAfterValue(const string& targetVal) {
    TwoListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr || targetNode->next == nullptr) {
        throw runtime_error("error");
    }
    
    TwoListNode* temp = targetNode->next;
    targetNode->next = temp->next;
    
    if (temp->next != nullptr) {
        temp->next->prev = targetNode;
    }
    
    if (temp == tail) {
        tail = targetNode;
    }
    
    delete temp;
}

void TwoList::removeBeforeValue(const string& targetVal) {
    TwoListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr || targetNode->prev == nullptr) {
        throw runtime_error("error");
    }
    
    if (targetNode->prev == head) {
        removeFromHead();
        return;
    }
    
    TwoListNode* temp = targetNode->prev;
    targetNode->prev = temp->prev;
    
    if (temp->prev != nullptr) {
        temp->prev->next = targetNode;
    }
    
    delete temp;
}

void TwoList::removeByValue(const string& targetVal) {
    TwoListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }
    
    if (targetNode == head) {
        removeFromHead();
        return;
    }
    
    if (targetNode == tail) {
        removeFromTail();
        return;
    }
    
    if (targetNode->prev != nullptr) {
        targetNode->prev->next = targetNode->next;
    }
    
    if (targetNode->next != nullptr) {
        targetNode->next->prev = targetNode->prev;
    }
    
    delete targetNode;
}

auto TwoList::getElementByValue(const string &targetVal) const -> string 
{
    TwoListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }
    return targetNode->data;
}

auto TwoList::findByVal(const string &val) const -> TwoListNode *
{
    TwoListNode* curr = head;    
    while (curr != nullptr) {
        if (curr->data == val) {
            return curr;
        }
        curr = curr->next;
    }   
    return nullptr;
}

void TwoList::print() const {
    if (head == nullptr) {
        cout << "error" << '\n';
        return;
    }    
    TwoListNode* curr = head;
    cout << "список: ";
    while (curr != nullptr) {
        cout << curr->data;
        if (curr->next != nullptr) {
            cout << " <-> ";
        }
        curr = curr->next;
    }
    cout << '\n';
}


void TwoList::saveToFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    TwoListNode* current = head;
    while (current != nullptr) {
        file << current->data << " ";
        current = current->next;
    }
}

void TwoList::loadFromFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    string value;
    while (file >> value) {
        addToTail(value);
    }
}

void TwoList::saveToBinaryFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    TwoListNode* current = head;
    while (current != nullptr) {
        size_t strSize = current->data.size();
        auto streamSize = static_cast<streamsize>(strSize);
        file.write(reinterpret_cast<const char*>(&strSize), sizeof(strSize));
        file.write(current->data.c_str(), streamSize);
        current = current->next;
    }
}

void TwoList::loadFromBinaryFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    while (file.peek() != EOF) {
        size_t strSize;
        file.read(reinterpret_cast<char*>(&strSize), sizeof(strSize));
        if (file.eof()){ 
            break;
        }
        auto streamSize = static_cast<streamsize>(strSize);
        char* buffer = new char[strSize + 1];
        file.read(buffer, streamSize);
        buffer[strSize] = '\0';
        addToTail(string(buffer));
        delete[] buffer;
    }
}

File: lab3/zad2/go/twolist/twolist.go
package twolist

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
)

type TwoList struct {
	head *TwoListNode
	tail *TwoListNode
}

func NewTwoList() *TwoList {
	return &TwoList{}
}

func (tl *TwoList) AddToHead(val string) {
	newNode := &TwoListNode{Data: val}
	if tl.head == nil {
		tl.head = newNode
		tl.tail = newNode
	} else {
		newNode.Next = tl.head
		tl.head.Prev = newNode
		tl.head = newNode
	}
}

func (tl *TwoList) AddToTail(val string) {
	newNode := &TwoListNode{Data: val}
	if tl.head == nil {
		tl.head = newNode
		tl.tail = newNode
	} else {
		newNode.Prev = tl.tail
		tl.tail.Next = newNode
		tl.tail = newNode
	}
}

func (tl *TwoList) AddAfterValue(targetVal, newVal string) error {
	targetNode := tl.findByVal(targetVal)
	if targetNode == nil {
		return errors.New("error")
	}

	newNode := &TwoListNode{Data: newVal}
	newNode.Next = targetNode.Next
	newNode.Prev = targetNode

	if targetNode.Next != nil {
		targetNode.Next.Prev = newNode
	}
	targetNode.Next = newNode

	if targetNode == tl.tail {
		tl.tail = newNode
	}
	return nil
}

func (tl *TwoList) AddBeforeValue(targetVal, newVal string) error {
	targetNode := tl.findByVal(targetVal)
	if targetNode == nil {
		return errors.New("error")
	}

	if targetNode == tl.head {
		tl.AddToHead(newVal)
		return nil
	}

	newNode := &TwoListNode{Data: newVal}
	newNode.Next = targetNode
	newNode.Prev = targetNode.Prev

	if targetNode.Prev != nil {
		targetNode.Prev.Next = newNode
	}
	targetNode.Prev = newNode
	return nil
}

func (tl *TwoList) RemoveFromHead() error {
	if tl.head == nil {
		return errors.New("error")
	}
	if tl.head == tl.tail {
		tl.head = nil
		tl.tail = nil
	} else {
		tl.head = tl.head.Next
		tl.head.Prev = nil
	}
	return nil
}

func (tl *TwoList) RemoveFromTail() error {
	if tl.head == nil {
		return errors.New("error")
	}

	if tl.head == tl.tail {
		tl.head = nil
		tl.tail = nil
	} else {
		tl.tail = tl.tail.Prev
		tl.tail.Next = nil
	}
	return nil
}

func (tl *TwoList) RemoveAfterValue(targetVal string) error {
	targetNode := tl.findByVal(targetVal)
	if targetNode == nil || targetNode.Next == nil {
		return errors.New("error")
	}

	temp := targetNode.Next
	targetNode.Next = temp.Next

	if temp.Next != nil {
		temp.Next.Prev = targetNode
	}

	if temp == tl.tail {
		tl.tail = targetNode
	}
	return nil
}

func (tl *TwoList) RemoveBeforeValue(targetVal string) error {
	targetNode := tl.findByVal(targetVal)
	if targetNode == nil || targetNode.Prev == nil {
		return errors.New("error")
	}

	if targetNode.Prev == tl.head {
		return tl.RemoveFromHead()
	}

	temp := targetNode.Prev
	targetNode.Prev = temp.Prev

	if temp.Prev != nil {
		temp.Prev.Next = targetNode
	}
	return nil
}

func (tl *TwoList) RemoveByValue(targetVal string) error {
	targetNode := tl.findByVal(targetVal)
	if targetNode == nil {
		return errors.New("error")
	}

	if targetNode == tl.head {
		return tl.RemoveFromHead()
	}

	if targetNode == tl.tail {
		return tl.RemoveFromTail()
	}

	if targetNode.Prev != nil {
		targetNode.Prev.Next = targetNode.Next
	}

	if targetNode.Next != nil {
		targetNode.Next.Prev = targetNode.Prev
	}
	return nil
}

func (tl *TwoList) GetElementByValue(targetVal string) (string, error) {
	targetNode := tl.findByVal(targetVal)
	if targetNode == nil {
		return "", errors.New("error")
	}
	return targetNode.Data, nil
}

func (tl *TwoList) findByVal(val string) *TwoListNode {
	curr := tl.head
	for curr != nil {
		if curr.Data == val {
			return curr
		}
		curr = curr.Next
	}
	return nil
}

func (tl *TwoList) Print() {
	if tl.head == nil {
		fmt.Println("error")
		return
	}

	curr := tl.head
	fmt.Print("список: ")
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Next != nil {
			fmt.Print(" <-> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func (tl *TwoList) SaveToFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := tl.head
	for curr != nil {
		_, err := file.WriteString(curr.Data + " ")
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (tl *TwoList) LoadFromFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	content, err := io.ReadAll(file)
	if err != nil {
		return err
	}

	values := strings.Fields(string(content))
	for _, val := range values {
		tl.AddToTail(val)
	}
	return nil
}

func (tl *TwoList) SaveToBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := tl.head
	for curr != nil {
		strSize := uint64(len(curr.Data))
		err := binary.Write(file, binary.LittleEndian, strSize)
		if err != nil {
			return err
		}
		_, err = file.WriteString(curr.Data)
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (tl *TwoList) LoadFromBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	for {
		var strSize uint64
		err := binary.Read(file, binary.LittleEndian, &strSize)
		if err != nil {
			if err == io.EOF {
				break
			}
			return err
		}

		buffer := make([]byte, strSize)
		_, err = file.Read(buffer)
		if err != nil {
			return err
		}
		tl.AddToTail(string(buffer))
	}
	return nil
}
